src/v/kafka/client/topic_cache.cc

```diff
@@ -18,29 +18,67 @@
 
 namespace kafka::client {
 
-void topic_cache::apply(
-  const chunked_vector<metadata_response::topic>& topics) {
-    topics_t new_cache;
-    new_cache.reserve(topics.size());
-    for (const auto& t : topics) {
-        static_assert(
-          api_version_for(metadata_request::api_type::key) < api_version(12),
-          "topic::name is nullable in v12+");
-        auto& cache_t = new_cache.emplace(*t.name, topic_data{}).first->second;
-        cache_t.authorized_operations = t.topic_authorized_operations;
-        if (!t.partitions.empty()) {
-            cache_t.replication_factor = t.partitions[0].replica_nodes.size();
+topic_cache::topic_data::topic_data(const metadata_response::topic& t)
+  : authorized_operations(t.topic_authorized_operations)
+  , last_update(ss::lowres_clock::now()) {
+    update_partitions(t.partitions);
+}
+
+void topic_cache::topic_data::update(const metadata_response::topic& t) {
+    authorized_operations = t.topic_authorized_operations;
+    last_update = ss::lowres_clock::now();
+    update_partitions(t.partitions);
+}
+
+void topic_cache::topic_data::update_partitions(
+  const chunked_vector<metadata_response::partition>& parts) {
+    bool rf_set = false;
+    if (partitions.size() < parts.size()) {
+        partitions.reserve(parts.size());
+    }
+    for (const auto& p : parts) {
+        if (p.error_code != error_code::none) {
+            // skip updating this partition
+            continue;
         }
-        cache_t.partitions.reserve(t.partitions.size());
-        for (const auto& p : t.partitions) {
-            cache_t.partitions.emplace(
+        const auto it = partitions.find(p.partition_index);
+        if (it != partitions.end()) {
+            if (it->second.leader_epoch > p.leader_epoch) {
+                // ignore stale update
+                continue;
+            }
+            it->second.leader = p.leader_id;
+            it->second.leader_epoch = p.leader_epoch;
+        } else {
```

Comment:
Note: This is safe, since partition count can't be reduced for an existing topic.

A lot of the changes in this commit seem unrelated to TTL expiry.